backend/layers/persistence/persistence.py

```diff
@@ -168,6 +171,18 @@ def _row_to_dataset_version(self, row: Any, canonical_dataset: CanonicalDataset,
             canonical_dataset,
         )
 
+    def _sort_datasets(self, row: Any, datasets: List[DatasetVersion]) -> List[DatasetVersion]:
+        """
+        Sorts datasets by the order they appear in the collection version's datasets array if custom
+        order is enabled, otherwise sorts datasets by cell count.
+        """
+        if row.datasets_custom_ordered:
+            datasets_order = [DatasetVersionId(str(id)) for id in row.datasets]
+            return sorted(datasets, key=lambda d: datasets_order.index(d.version_id))
+        return sorted(
+            datasets, key=lambda d: 0 if d is None or d.metadata is None else d.metadata.cell_count, reverse=True
```

Comment:
in what scenario is `d` going to evaluate to `None`?

Reply:
Looks like this is no longer possible; I have removed `is None`.